### PR TITLE
Create Modal Manager

### PR DESCRIPTION
**Pull request by jbuckner**
This adds a customizable modal manager.

It is based off of the donation modals we currently have in Petabox with formalized configuration and expanded capabilities like custom content.

## Usage
```html
<!-- index.html -->
<script type="module">
  import '@internetarchive/modal-manager';
  import { ModalConfig } from '@internetarchive/modal-manager';
</script>

<style>
  /* add the following styles to ensure proper modal visibility */
    body.modal-manager-open {
      overflow: hidden;
    }

    modal-manager {
      display: none;
    }

    modal-manager[mode='open'] {
      display: block;
    }
</style>

<modal-manager></modal-manager>

<script>
  // show a simple modal
  const manager = document.querySelector('modal-manager');
  const config = new ModalConfig();
  config.headline = 'Hi, Everybody!';
  config.message = 'Hi, Doctor Nick!';
  manager.showModal(config)

  // to hide the modal call `closeModal()`:
  manager.closeModal();
</script>
```

## Basic Modal
modal-screenshot

## Error Modal
Screen Shot 2020-06-22 at 2 10 37 PM

## Processing Modal
Screen Shot 2020-06-22 at 2 11 29 PM

## Custom Content
Screen Shot 2020-06-22 at 2 10 55 PM

### CODE REVIEW

File: src/modal-header-colors.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Some some colors that can be used for the header
+ *
+ * @export
+ * @enum {number}
+ */
+export enum ModalHeaderColors {
```
Comment (shaneriley) [inline src/modal-header-colors.ts:7]: I have yet to write TypeScript, and am wondering why an enum of strings is preferred to a singleton object with the same property assignments.

File: src/modal-config.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+   * @memberof ModalConfig
+   */
+  message?: TemplateResult;
+
+  /**
+   * The color of the header
+   *
+   * @memberof ModalConfig
+   */
+  headerColor = '#36A483';
```
Comment (shaneriley) [inline src/modal-config.ts:43]: **Suggestion**: move this to `ModalHeaderColors`

File: src/modal-template.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+  render(): TemplateResult {
+    return html`
+    <div class="modal-wrapper">
+      <div class="head-spacer"></div>
+      <div class="modal-container">
+        <header style="background-color: ${this.config.headerColor}">
+          ${this.config.allowUserToClose ? this.closeButtonTemplate : ''}
+          <div class="logo-icon">
+            <img src="https://archive.org/images/logo_arches.png">
```
Comment (shaneriley) [inline src/modal-template.ts:29]: For a11y, add an alt attribute describing the logo

File: src/modal-template.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+          </div>
+          <h1>${this.config.title}</h1>
+          <h2>${this.config.subtitle}</h2>
+        </header>
+        <div class="body">
```
Comment (shaneriley) [inline src/modal-template.ts:34]: **Suggestion**: rename this class name to avoid using the same name as a native HTML element.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export { ModalManager, ModalManagerMode, ModalManagerInterface } from './lib/modal-manager.d';
```
[the remark this answers is not in the document]
Comment (jbuckner) [reply index.d.ts:1]: With Typescript, in addition to the main `index.js` exports file, you create a declarations file `index.d.ts` that defines all of the types that you're using. This is useful when Typescript consumers pull your package in. They're just getting the javascript installed, but the declaration file lets them know the structure of the javascript without needing all the source Typescript files.

In this case, `ModalManagerMode` (an `enum`) and `ModalManagerInterface` (an `interface`) are not supported by javascript so you put them here so Typescript consumers know about them while leaving the Javascript side alone.

File: src/assets/close-icon.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { html } from 'lit-element';
```
[the remark this answers is not in the document]
Comment (jbuckner) [reply src/assets/close-icon.ts:1]: Yeah I was thinking about that but my concern is that I'm going to be importing an entire library of icons (and the file size that comes with it) for just this one close icon.

File: src/modal-header-colors.ts (hunk 1 above)
Comment (jbuckner) [reply src/modal-header-colors.ts:7]: So this is actually not a great use case for an enum and I was considering removing it from this repo.. it's really just a few presets that we can use for the header colors. We're not restricting the `headerColor` that you can pass in since it's just a string in `ModalConfig.headerColor`.

Regarding your question about enum vs singleton object.. a better use case of an enum would be if we only allowed specific colored modals. The configuration would instead look like this:

```
class ModalConfig {
  headerColor: ModalHeaderColors = ModalHeaderColors.Green; // default value
}
...
const config = new ModalConfig();
config.headerColor = ModalHeaderColors.Blue;
config.headerColor = ModalHeaderColors.Purple; // compile error
```

Now we've restricted our options of what we are allowing `headerColor` to be. The Typescript compiler will now enforce the developer to only allow options that are in `ModalHeaderColors`.

Enums are really useful when you have a fixed set of options and you want to enforce developers to only use one of those options.

File: src/assets/close-icon.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (shaneriley) [reply src/assets/close-icon.ts:1]: I figured you'd pull the individual file directly, since you don't need a customElement for it. You should be able to install the package within this project, then import https://github.com/internetarchive/iaux/blob/master/packages/ia-icons/src/icons/close.js

File: src/modal-header-colors.ts (hunk 1 above)
Comment (shaneriley) [reply src/modal-header-colors.ts:7]: I figured it had to do with enforcing one of a set of inputs. Does the same value enforcement occur when the enum is populated with integer values?

File: src/modal-manager.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+export class ModalManager extends LitElement implements ModalManagerInterface {
+  /**
+   * The current mode of the ModalManager
+   *
+   * Current options are `modal` or `closed`
+   *
+   * @type {ModalManagerMode}
+   * @memberof ModalManager
+   */
+  @property({ type: String, reflect: true }) mode: ModalManagerMode = ModalManagerMode.Closed;
```
Comment (iisa) [inline src/modal-manager.ts:52]: q: what does `reflect` mean?

File: src/modal-template.ts (hunk 4 above)
Comment (iisa) [reply src/modal-template.ts:34]: do we want to go further and semantically change the dom node? section, main, perhaps? - to pair with the `<header>` tag above.

File: src/assets/close-icon.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (iisa) [reply src/assets/close-icon.ts:1]: 🤔 💭  can icons be a mono-repo with an option to import all and 1x1? is this our case for lerna? 😄 💭

[the remark this answers is not in the document]
Comment (shaneriley) [reply src/assets/close-icon.ts:1]: https://codesandbox.io/s/eager-sunset-v9cs7?file=/src/index.js

[the remark this answers is not in the document]
Comment (jbuckner) [reply src/assets/close-icon.ts:1]: But by adding `ia-icons` as an npm dependency, you're inherently bringing in the entire npm module with it, even though you're only using that one icon. When Pika packages the module, it creates a single file containing everything included in it, not just the imported files.

That's a lot of file-size overhead (that will continue to grow) for a single icon.

[the remark this answers is not in the document]
Comment (shaneriley) [reply src/assets/close-icon.ts:1]: The what is the solution? To not have an icons package? To have many identical copies of the same asset that, when design updates it, have to all be updated manually, leading to forgotten uses of the asset? This icon package will be included in the Petabox dependencies when ia-topnav switches to use it, and the plan was to also have the new BookReader chrome use it for the side menu icons. Since they're already in the codebase, there won't be any additional overhead when this is integrated.

[the remark this answers is not in the document]
Comment (jbuckner) [reply src/assets/close-icon.ts:1]: I like Isa's idea of a lerna monorepo of icons that can each be published individually and put the main `ia-icon` component in there that imports all of them. Then our components can pull them in piecemeal and we'll still have a central repo of all the icons.

File: src/modal-header-colors.ts (hunk 1 above)
Comment (jbuckner) [reply src/modal-header-colors.ts:7]: Yeah. Actually enums are integers by default, but assigning string values to them turns them into a string. There's actually [quite a bit](https://www.typescriptlang.org/docs/handbook/enums.html) you can do with enums, even setting the values to callbacks.

File: src/modal-manager.ts (hunk 7 above)
Comment (jbuckner) [reply src/modal-manager.ts:52]: `reflect` causes the property to update the element tag attributes. Normally if a property changes its value inside the component, the tag stays the same. It's a one-way street, the consumer sets it and it's used inside. With reflect, when the property value changes, it updates the tag too. ie:
```
<modal-manager mode='closed'></modal-manager>
...
modalManager.showModal(...) // causes mode to change to `modal`
```
Without `reflect`, the tag would not change. When you enable `reflect`, the tag would become:
```
<modal-manager mode='modal'></modal-manager>
```

The reason we're using `reflect` here is because the modal visibility is largely just controlled by CSS, but the functionality of triggering that open/close is based on internal logic. The consumer of the modal needs to add these styles:
```
  modal-manager {
    position: absolute;
    width: 100%;
    height: 100%;
    z-index: 250;
  }

  modal-manager[mode='closed'] { <--- these are the reflected properties
    display: none;
  }

  modal-manager[mode='modal'] { <--- these are the reflected properties
    display: block;
  }
```

File: src/assets/close-icon.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (iisa) [reply src/assets/close-icon.ts:1]: let's do it!

File: src/modal-manager.ts (hunk 7 above)
Comment (iisa) [reply src/modal-manager.ts:52]: thanks for the break down.

File: src/assets/close-icon.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (shaneriley) [reply src/assets/close-icon.ts:1]: So the solution is to have each SVG file be its own NPM package? That feels like overkill.

[the remark this answers is not in the document]
Comment (jbuckner) [reply src/assets/close-icon.ts:1]: I think they'd only be like 2 files each: `index.js` and `package.json`. We don't need tests or anything like that and since it'd be in lerna, we'd just have to make one good one then duplicate the directory and push the new icon.

[the remark this answers is not in the document]
Comment (jbuckner) [reply src/assets/close-icon.ts]: Now imported from iaux-icons

File: src/modal-template.ts (hunk 4 above)
Comment (jbuckner) [reply src/modal-template.ts]: I switched this to use `<section>`

File: package.json [hunk 8]
```diff
@@ -0,0 +1,74 @@
+{
+  "name": "@internetarchive/modal-manager",
+  "version": "0.0.0-alpha.19",
+  "description": "A Modal Manager Web Component",
+  "author": "modal-manager",
```
Comment (shaneriley) [inline package.json:5]: I usually just drop the author field since they're released from the team account.

File: src/modal-config.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+ * Configuration to show a modal
+ *
+ * @export
+ * @class ModalConfig
+ */
+export class ModalConfig {
+  /**
+   * The title that shows in the header
+   *
+   * @type {(TemplateResult | undefined)}
```
Comment (shaneriley) [inline src/modal-config.ts:13]: Wouldn't this be `TemplateResult | LitHtml object`?

File: src/modal-manager-controller.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+   * @returns {void}
+   * @memberof ModalManagerController
+   */
+  closeModal(): void {
+    return this.modalManager.closeModal();
+  }
+
+  /** @inheritdoc */
+  firstUpdated(): void {
+    window.addEventListener('resize', () => {
```
Comment (shaneriley) [inline src/modal-manager-controller.ts:90]: This should be throttled so there are far fewer calls while drag-resizing.

Additionally, is there a teardown method that removes this listener?

File: src/modal-manager-controller.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+   *
+   * @private
+   * @param {CustomEvent} e
+   * @memberof ModalManagerController
+   */
+  private modalModeChanged(e: CustomEvent): void {
+    const mode = e.detail.mode as ModalManagerMode;
+    switch (mode) {
+      case ModalManagerMode.Modal:
+        document.body.style.overflow = 'hidden';
```
Comment (shaneriley) [inline src/modal-manager-controller.ts:116]: **Suggestion**: toggle a class on the body rather than directly set a CSS property, especially since the default value for overflow is "visible" rather than "auto". Then the parent document can control what happens in these two states.

File: src/modal-manager.ts [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+   * Show a modal from a given ModalConfig
+   *
+   * @param config ModalConfig
+   * @param customModalContent TemplateResult | undefined
+   * @param userClosedModalCallback () => void | undefined an optional callback when the modal is closed
+   */
+  showModal(options: {
+    config: ModalConfig;
+    customModalContent?: TemplateResult;
+    userClosedModalCallback?: () => void;
```
Comment (shaneriley) [inline src/modal-manager.ts:44]: Why are we using a callback instead of emitting a custom event on close? That should prevent the infinite loop as well.

File: src/modal-manager.ts [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+  }
+
+  /**
+   * Called when the backdrop is clicked
+   *
+   * @private
+   * @memberof ModalManager
+   */
+  private backdropClicked(): void {
+    if (this.closeOnBackdropClick) {
```
Comment (shaneriley) [inline src/modal-manager.ts:164]: You didn't need an `/* istanbul ignore else */` here as well?

File: src/modal-template.ts [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+    `;
+  }
+
+  /** @inheritdoc */
+  static get styles(): CSSResult {
+    const processingImageSize = css`var(--processingImageSize, 75px)`;
+
+    const modalCornerRadius = css`var(--modalCornerRadius, 10px)`;
+    const modalBorder = css`var(--modalBorder, 2px solid black)`;
+    // if the content of the modal is too big to fit on screen, this sets the bottom margin
```
Comment (shaneriley) [inline src/modal-template.ts:112]: You could alternately set a `max-height` of something like `calc(100vh - 75px)` and set `overflow: auto` to allow the contents to scroll.

File: src/modal-template.ts [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+    // it's not exact, but a close estimation
+    const modalBottomMarginCss = css`var(--modalBottomMargin, 25px)`;
+    const modalTopMarginCss = css`var(--modalTopMargin, 50px)`;
+    const modalHeaderBottomPaddingCss = css`var(--modalHeaderBottomPadding, 0.5em)`;
+
+    const modalBottomPadding = css`var(--modalBottomPadding, 20px)`;
+    const scrollOffset = css`var(--modalScrollOffset, 5px)`;
+
+    const titleFontSize = css`var(--modalTitleFontSize, 18px)`;
```
Comment (shaneriley) [inline src/modal-template.ts:121]: We've been using rem units for font sizes so they scale proportionally when the root element's font size changes. Like with Petabox, I like to start with the html element set to 10px, then all font sizes in rem are easy to determine. Then if we want to alter all font sizes proportionally based on viewport size, for example, we're all set.

File: custom-elements.json [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+          "type": "Number",
+          "description": "An action counter",
+          "default": 0
+        }
+      ],
+      "events": [],
+      "slots": [],
+      "cssProperties": [
+        {
+          "name": "--modal-manager-text-color",
```
Comment (shaneriley) [inline custom-elements.json:25]: I don't see this custom variable used in the component code. Looks like there are also a lot of CSS variables that could be defined here and/or in the readme.

File: src/modal-config.ts (hunk 9 above)
Comment (jbuckner) [reply src/modal-config.ts:13]: `TemplateResult` is the type that gets output when you use ``` html`Foo <b>Bar</b>` ``` so in Typescript, the `render()` method looks like this:
```
render(): TemplateResult {
  return html`<div> ... </div>`;
}
```

Technically, you can also pass in strings here, but the `TemplateResult` lets you put markup in the field.

File: src/modal-manager-controller.ts (hunk 11 above)
Comment (jbuckner) [reply src/modal-manager-controller.ts:116]: This is where things were getting tricky regarding responsibilities between the host and the modal manager itself.

Originally I had all of this logic in petabox so as to let it handle things like overflow, but there's a decent amount of interaction that has to happen for it to all work properly that I didn't want to saddle on the host.

I instead created this `<modal-manager-controller>`, which is a somewhat generic bridge between the host and the actual `<modal-manager>`, which it wraps and passes through. It makes a few assumptions like this one to require as little integration work on the consumer as possible.

Consumers could take the `<modal-manager>` directly and handle this logic without using the controller or even subclass the controller to handle their environment.

I might be able to just pull out a few of the environmental-specific things like this. I need to think of how to do that.

File: src/modal-config.ts (hunk 9 above)
Comment (shaneriley) [reply src/modal-config.ts:13]: Ah, ok. So that's the former object type. But if you're setting it to a default TemplateResult value, shouldn't the doc line be `@type {TemplateResult}`?

File: src/modal-manager.ts (hunk 12 above)
Comment (jbuckner) [reply src/modal-manager.ts:44]: This is a momentary action. We show the modal, then some action happens on it that will eventually result in it closing.

To me, event listeners are for long-lived things. Like internally, the `modal-manager-template` emits custom events that the `modal-manager` listens for.

I think using event listeners for something like this, the consumer will end up having to deal with making sure they bind and unbind every time the various actions happen so you don't respond to events that you didn't instantiate.

File: src/modal-config.ts (hunk 9 above)
Comment (jbuckner) [reply src/modal-config.ts:13]: Yeah I set a default value, but this lets people set it to `undefined` if they don't want the default.

File: src/modal-template.ts (hunk 14 above)
Comment (jbuckner) [reply src/modal-template.ts:112]: So this is interesting.. the `calc(100vh - 75px)` calculates the initial size correctly, but then doesn't change after that so if you resize your window, the height doesn't change.

For some reason that `window.onresize` event listener that is currently setting the `--containerHeight` causes the `calc()` to recalculate, even if I set any random CSS variable. If I remove that `onresize` the `calc()` stops working.

Comment (shaneriley) [reply src/modal-template.ts:112]: That's...odd. I thought any value with a relative measurement would recalculate with a repaint pass, like when the viewport resizes. Maybe it's calculating a pixel value because of the mixed units?

Comment (jbuckner) [reply src/modal-template.ts:112]: I'm not sure if this would have any effect, but since this is happening inside the shadow root, do you know if that affects these calculations? That was actually the original reason I put in that `window.onresize` ... I couldn't figure out a way to properly calculate the content's frame since it didn't have access to anything above its own DOM.

Comment (shaneriley) [reply src/modal-template.ts:112]: `window` is the same in shadow and light DOM, so I would imagine the same issue should present itself in a plain static page.

Comment (shaneriley) [reply src/modal-template.ts:112]: https://output.jsbin.com/mekezup behaves as expected, so maybe it is a shadow DOM issue.

Comment (jbuckner) [reply src/modal-template.ts:112]: https://output.jsbin.com/wenoquh/1 .. I took the same code you used, put it inside the shadowDOM and I'm seeing similar behavior where it renders properly at first, but then doesn't resize.

Comment (jbuckner) [reply src/modal-template.ts:112]: Sooo this looks like it's only an issue in Safari... I tested that in Firefox and Chrome and both were resizing properly

File: README.md [hunk 17]
```diff
@@ -1,2 +1,161 @@
...
+<style>
+  /* add the following styles to ensure proper modal visibility */
+  modal-manager {
+    position: absolute;
+    width: 100%;
+    height: 100%;
+    z-index: 250;
+  }
+
+  modal-manager[mode='closed'] {
```
[the remark this answers is not in the document]
Comment (jbuckner) [reply README.md:31]: Not particularly. It defaults to `closed` so this is basically the default state, but I removed the `mode` attribute selector so it defaults to `display: none`

File: src/modal-manager-controller.ts (hunk 10 above)
Comment (jbuckner) [reply src/modal-manager-controller.ts]: This is throttled now.

File: src/modal-template.ts (hunk 14 above)
Comment (jbuckner) [reply src/modal-template.ts:112]: I changed it to use the `calc(100vh .. )` calculation and that works nicely. In Firefox and Chrome, it updates as you resize. In Safari, there's a 100ms throttler to trigger the re-render.

File: src/modal-manager-controller.ts (hunk 11 above)
Comment (jbuckner) [reply src/modal-manager-controller.ts]: I switched it to use a class so the consumer just has to add
```
body.modal-manager-open {
  overflow: hidden;
}
```
and it'll add/remove that selector.

File: src/modal-config.ts (hunk 9 above)
Comment (jbuckner) [reply src/modal-config.ts:13]: I removed the default value.

File: src/modal-manager-controller.ts (hunk 11 above)
Comment (jbuckner) [reply src/modal-manager-controller.ts]: So I ended up dropping the `modal-manager-controller` class in favor of a sidecar class called the `ModalManagerHostBridge`. It is responsible for handling these host interactions and we provide a default class, which can be overridden by the consumer if they want to do things differently. It also avoids the awkward wrapping that the controller was causing.

File: src/modal-template.ts (hunk 15 above)
Comment (jbuckner) [reply src/modal-template.ts:121]: I tried setting these to use `rem` values, but they don't seem to change depending on the base font size. They always seem to be relative to the base font size of that type. ie `titleFontSize` is on an `h1`, which has a browser default of `16px`. Setting this to `1.4rem` became around `22px` instead of, say, `14px` for a `10px` base. I'm not sure if I'm setting that correctly or if that's intentional with the Shadow DOM.

Comment (shaneriley) [reply src/modal-template.ts:121]: In your demo HTML file, you need to set the font size to 10px.

```css
html {
  font-size: 10px;
}
```

The topnav does this to ensure parity between demo and Petabox integration.

Comment (jbuckner) [reply src/modal-template.ts]: Ah I had it set on the `body`. Now it's working.

File: custom-elements.json (hunk 16 above)
Comment (jbuckner) [reply custom-elements.json]: I updated this with the element details.

File: src/modal-manager.ts (hunk 13 above)
Comment (jbuckner) [reply src/modal-manager.ts:183]: Doesn't seem to need it
